### ISSUE TITLE
-Wignored-attributes in drivers/net/ethernet/intel/iavf/iavf_osdep.h

### ISSUE THREAD

**Issue by tpimh** [ClangBuiltLinux/linux#111]
```
drivers/net/ethernet/intel/i40evf/i40e_osdep.h:49:18: warning: 'format' attribute argument not supported: gnu_printf [-Wignored-attributes]
        __attribute__ ((format(gnu_printf, 3, 4)));
                        ^
1 warning generated.
```

**Comment by nickdesaulniers** [comment]
Probably want to use @ojeda 's proper attribute detection patches, once they land in -next.

**Comment by ojeda** [comment]
They have now landed in linux-next.

**Comment by ojeda** [comment]
The compiler attributes series were [merged into mainline](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=e468f5c06b5ebef3f6f3c187e51aa6daab667e57) yesterday, so we should get them for v4.20-rc1.

**Comment by nathanchance** [comment]
This is still an issue, we just can fix it now that we have the attributes tree merged.

**Comment by nickdesaulniers** [comment]
Is this in -next?

**Comment by nathanchance** [comment]
The warning?

**Comment by nickdesaulniers** [comment]
drivers/net/ethernet/intel/i40evf/i40e_osdep.h doesn't exist in torvalds/linux.

**Comment by nathanchance** [comment]
Got shuffled to `drivers/net/ethernet/intel/iavf/i40e_osdep.h` in commit 8062b2263a9f ("intel-ethernet: rename i40evf to iavf") then to `drivers/net/ethernet/intel/iavf/iavf_osdep.h` in commit 66bc8e0f591a ("iavf: finish renaming files to iavf").

**Comment by nathanchance** [comment]
@ojeda how would you suggesting adding this to the compiler attributes file? Clang obviously supports `__attribute__((__format__(...)))`, it just doesn't support `__attribute__((__format__(gnu_printf, ...)))`

I can only think of adding it based on `CONFIG_CC_IS_GCC` like so:

```diff
diff --git a/drivers/net/ethernet/intel/iavf/iavf_osdep.h b/drivers/net/ethernet/intel/iavf/iavf_osdep.h
index e6e0b0328706..a1cfabdd452c 100644
--- a/drivers/net/ethernet/intel/iavf/iavf_osdep.h
+++ b/drivers/net/ethernet/intel/iavf/iavf_osdep.h
@@ -46,7 +46,7 @@ struct iavf_virt_mem {

 #define iavf_debug(h, m, s, ...)  iavf_debug_d(h, m, s, ##__VA_ARGS__)
 extern void iavf_debug_d(void *hw, u32 mask, char *fmt_str, ...)
-       __attribute__ ((format(gnu_printf, 3, 4)));
+       __gnu_printf(3, 4);

 typedef enum iavf_status_code iavf_status;
 #endif /* _IAVF_OSDEP_H_ */
diff --git a/include/linux/compiler_attributes.h b/include/linux/compiler_attributes.h
index 19f32b0c29af..683c67a6c100 100644
--- a/include/linux/compiler_attributes.h
+++ b/include/linux/compiler_attributes.h
@@ -142,6 +142,11 @@
  */
 #define __printf(a, b)                  __attribute__((__format__(printf, a, b)))
 #define __scanf(a, b)                   __attribute__((__format__(scanf, a, b)))
+#ifdef CONFIG_CC_IS_GCC
+# define __gnu_printf(a, b)             __attribute__((__format__(gnu_printf, a, b)))
+#else
+# define __gnu_printf(a, b)
+#endif

 /*
  *   gcc: https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-gnu_005finline-function-attribute
```

Any guidance would be appreciated :)

**Comment by nathanchance** [comment]
As it turns out, GCC has the `printf` attribute aliased to `gnu_printf` since at least 2010 ([source](https://gcc.gnu.org/git/?p=gcc.git;a=blob;f=gcc/c-family/c-format.c;h=327fa29937d2a7e0a58438e5468b64e91f10c83f;hb=HEAD#l4077)):

```c
/* Attributes such as "printf" are equivalent to those such as
   "gnu_printf" unless this is overridden by a target.  */
static const target_ovr_attr gnu_target_overrides_format_attributes[] =
{
  { "gnu_printf",   "printf" },
  { "gnu_scanf",    "scanf" },
  { "gnu_strftime", "strftime" },
  { "gnu_strfmon",  "strfmon" },
  { NULL,           NULL }
};
```

The formats only get overridden on Windows (so who cares amirite) so it should be totally fine to convert that to `__printf`. I'll draft up the commit now, thank you for looking into that @ojeda :)

**Comment by ojeda** [comment]
You're welcome! The wording seems fine :) Thanks for the patch!

**Comment by nickdesaulniers** [comment]
@nathanchance what happened to that patch? Broken link

**Comment by nathanchance** [comment]
The dev-queue branch gets constantly rebased until it gets sent to net-next it seems.

Current hash: https://git.kernel.org/jkirsher/next-queue/c/9e5f261679e81a027bb33efbbb2720aac80e2fe7

**Comment by nickdesaulniers** [comment]
link to the tree instead: https://git.kernel.org/pub/scm/linux/kernel/git/jkirsher/next-queue.git/log/?h=dev-queue

**Comment by nathanchance** [comment]
https://git.kernel.org/pub/scm/linux/kernel/git/davem/net-next.git/commit/?id=3dc215416673f9ba8f136aafbb0069ed9ebff376

Finally this has been picked up into net-next for 5.3!

**Comment by nathanchance** [comment]
Merged into mainline: https://git.kernel.org/torvalds/c/3dc215416673f9ba8f136aafbb0069ed9ebff376
